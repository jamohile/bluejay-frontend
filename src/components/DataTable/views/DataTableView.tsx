import * as React from 'react';
import { DataRow } from "../../../utilities/store/reducers/data";
import './dataTableView.css';

/**
 * Interfaces, used to allow typing in all views.
 */
interface DataTableViewProps {
    rows: DataRow[],
    loaded: boolean,
    loading: boolean,
    error: boolean,
    excessRows: number,
    onReload: () => void
}

interface DataTableViewState {
    stale: boolean;
}

interface ReloadViewProps {
    onReload: () => void;
    stale: boolean
}

interface ExcessViewProps {
    num: number;
}
interface RowViewProps {
    number: number,
    value: string,
    count: number,
    age: number
}

interface RowsViewProps {
    rows: DataRow[];
    excessRows: number;
}

/**
 * Uses children Views to display a table of the rows coresponding to the selected field.
 * Not using a table was chosen for the sake of formatting in this case.
 * If necessary, implementing this using existing table views could be done.
 */
export class DataTableView extends React.Component<DataTableViewProps, DataTableViewState> {
    constructor(props: DataTableViewProps) {
        super(props);
        this.state = {
            stale: false
        }
    }

    componentWillReceiveProps() {
        this.setState({ stale: false })
        setTimeout(() => this.setState({ stale: true }), 10000);
    }
    render() {
        const { rows, loaded, loading, error, excessRows, onReload } = this.props;

        return (
            <div className='dataTable'>
                <HeaderView />
                {
                    error &&
                    <ErrorView />
                }

                {
                    loading &&
                    <LoadingView />
                }
                {
                    !loading && (loaded || error) &&
                    <ReloadButton
                        onReload={onReload}
                        stale={this.state.stale}
                    />
                }
                {
                    !error && loaded && !loading &&
                    <RowsView
                        rows={rows}
                        excessRows={excessRows}
                    />
                }
            </div>
        )
    }
}

// The header of the DataTable with column labels.
export const HeaderView = () => (
    <div className='headerView'>
        <span className='text number'>#</span>
        <span className='text value'>Value</span>
        <span className='text count'>Count</span>
        <span className='text age'>Age</span>
    </div>
)

// An animated loading indicator, shown when loading data for the given field.
export const LoadingView = () => (
    <div className='cardView loadingView'>
        <span className='text'>One moment...</span>
    </div>
)

// A button to trigger reload of the current field. If stale, reccomends a reload.
export const ReloadButton = ({ onReload, stale }: ReloadViewProps) => (
    <div className='cardView reloadView'
        onClick={onReload}
    >
        <span className='text'>Reload <b>{stale ? ' - This data may be out of date.' : ''}</b></span>
    </div>
)

// Displays an error if it occurred.
export const ErrorView = () => (
    <div className='cardView errorView'>
        <span className='text'>Something went wrong.</span>
    </div>
)

// Displays the number of excess columns.
export const ExcessView = ({ num }: ExcessViewProps) => (
    <div className='cardView excessView'>
        <span className='text'>{num} rows were omitted.</span>
    </div>
)

// Displays data for a single row of data.
export const RowView = ({ number, value, count, age }: RowViewProps) => (
    <div className='cardView rowView'>
        <span className='faded text number'>{number}</span>
        <span className='text value'>{value}</span>
        <span className='faded text count'>{count}</span>
        <span className='faded text age'>{age}</span>
    </div>
)

// Takes in a set of rows, displays each one along with an ExcessView if needed.
export const RowsView = ({ rows, excessRows }: RowsViewProps) => (
    <div className='rowsView'>
        {
            excessRows > 0 &&
            <ExcessView num={excessRows} />
        }
        {
            rows.map((r: DataRow, i: number) => (
                <RowView
                    key={'row_' + i}
                    number={i + 1}
                    value={r.value}
                    count={r.count}
                    age={r.age}
                />
            ))
        }
    </div>
)