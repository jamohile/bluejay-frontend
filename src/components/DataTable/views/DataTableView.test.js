import React from 'react';
import renderer from 'react-test-renderer';
import { HeaderView, LoadingView, ReloadButton, ExcessView, RowView, RowsView, DataTableView } from './DataTableView';

const sampleRows = [
    {
        value: 'apple',
        age: 4,
        count: 7
    }, {
        value: 'banana',
        age: 100,
        count: 155
    }, {
        value: 'pear',
        age: 5,
        count: 17
    }];

    it('Renders HeaderView.', () => {
        const tree = renderer.create(
            <HeaderView/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Renders LoadingView.', () => {
        const tree = renderer.create(
            <LoadingView/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders normal ReloadView.', () => {
        const tree = renderer.create(
            <ReloadButton/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders stale ReloadView.', () => {
        const tree = renderer.create(
            <ReloadButton stale/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders ExcessView.', () => {
        const tree = renderer.create(
            <ExcessView num/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders a single RowView', () => {
        const row = sampleRows[0];
        const tree = renderer.create(
            <RowView
                age = {row.age}
                value = {row.value}
                count = {row.count}
                key = {row.value}
                number = {0}
            />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders RowsView', () => {
        const tree = renderer.create(
           <RowsView
                rows={sampleRows}

           />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders DataTable with excess', () => {
        const tree = renderer.create(
           <DataTableView
                rows={sampleRows}
                excessRows={123}
                onReload={() => {}}
                loaded
           />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders DataTable with excess', () => {
        const tree = renderer.create(
           <DataTableView
                rows={sampleRows}
                excessRows={123}
                onReload={() => {}}
                loaded
           />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders Loading DataTable', () => {
        const tree = renderer.create(
           <DataTableView
                rows={sampleRows}
                excessRows={123}
                onReload={() => {}}
                loading
           />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders error DataTable', () => {
        const tree = renderer.create(
           <DataTableView
                rows={sampleRows}
                excessRows={123}
                onReload={() => {}}
                error
           />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });