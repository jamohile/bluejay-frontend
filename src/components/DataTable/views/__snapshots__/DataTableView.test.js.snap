// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders DataTable with excess 1`] = `
<div
  className="dataTable"
>
  <div
    className="headerView"
  >
    <span
      className="text number"
    >
      #
    </span>
    <span
      className="text value"
    >
      Value
    </span>
    <span
      className="text count"
    >
      Count
    </span>
    <span
      className="text age"
    >
      Age
    </span>
  </div>
  <div
    className="cardView reloadView"
    onClick={[Function]}
  >
    <span
      className="text"
    >
      Reload 
      <b>
        
      </b>
    </span>
  </div>
  <div
    className="rowsView"
  >
    <div
      className="cardView excessView"
    >
      <span
        className="text"
      >
        123
         rows were omitted.
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        1
      </span>
      <span
        className="text value"
      >
        apple
      </span>
      <span
        className="faded text count"
      >
        7
      </span>
      <span
        className="faded text age"
      >
        4
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        2
      </span>
      <span
        className="text value"
      >
        banana
      </span>
      <span
        className="faded text count"
      >
        155
      </span>
      <span
        className="faded text age"
      >
        100
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        3
      </span>
      <span
        className="text value"
      >
        pear
      </span>
      <span
        className="faded text count"
      >
        17
      </span>
      <span
        className="faded text age"
      >
        5
      </span>
    </div>
  </div>
</div>
`;

exports[`Renders DataTable with excess 2`] = `
<div
  className="dataTable"
>
  <div
    className="headerView"
  >
    <span
      className="text number"
    >
      #
    </span>
    <span
      className="text value"
    >
      Value
    </span>
    <span
      className="text count"
    >
      Count
    </span>
    <span
      className="text age"
    >
      Age
    </span>
  </div>
  <div
    className="cardView reloadView"
    onClick={[Function]}
  >
    <span
      className="text"
    >
      Reload 
      <b>
        
      </b>
    </span>
  </div>
  <div
    className="rowsView"
  >
    <div
      className="cardView excessView"
    >
      <span
        className="text"
      >
        123
         rows were omitted.
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        1
      </span>
      <span
        className="text value"
      >
        apple
      </span>
      <span
        className="faded text count"
      >
        7
      </span>
      <span
        className="faded text age"
      >
        4
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        2
      </span>
      <span
        className="text value"
      >
        banana
      </span>
      <span
        className="faded text count"
      >
        155
      </span>
      <span
        className="faded text age"
      >
        100
      </span>
    </div>
    <div
      className="cardView rowView"
    >
      <span
        className="faded text number"
      >
        3
      </span>
      <span
        className="text value"
      >
        pear
      </span>
      <span
        className="faded text count"
      >
        17
      </span>
      <span
        className="faded text age"
      >
        5
      </span>
    </div>
  </div>
</div>
`;

exports[`Renders ExcessView. 1`] = `
<div
  className="cardView excessView"
>
  <span
    className="text"
  >
     rows were omitted.
  </span>
</div>
`;

exports[`Renders HeaderView. 1`] = `
<div
  className="headerView"
>
  <span
    className="text number"
  >
    #
  </span>
  <span
    className="text value"
  >
    Value
  </span>
  <span
    className="text count"
  >
    Count
  </span>
  <span
    className="text age"
  >
    Age
  </span>
</div>
`;

exports[`Renders Loading DataTable 1`] = `
<div
  className="dataTable"
>
  <div
    className="headerView"
  >
    <span
      className="text number"
    >
      #
    </span>
    <span
      className="text value"
    >
      Value
    </span>
    <span
      className="text count"
    >
      Count
    </span>
    <span
      className="text age"
    >
      Age
    </span>
  </div>
  <div
    className="cardView loadingView"
  >
    <span
      className="text"
    >
      One moment...
    </span>
  </div>
</div>
`;

exports[`Renders LoadingView. 1`] = `
<div
  className="cardView loadingView"
>
  <span
    className="text"
  >
    One moment...
  </span>
</div>
`;

exports[`Renders RowsView 1`] = `
<div
  className="rowsView"
>
  <div
    className="cardView rowView"
  >
    <span
      className="faded text number"
    >
      1
    </span>
    <span
      className="text value"
    >
      apple
    </span>
    <span
      className="faded text count"
    >
      7
    </span>
    <span
      className="faded text age"
    >
      4
    </span>
  </div>
  <div
    className="cardView rowView"
  >
    <span
      className="faded text number"
    >
      2
    </span>
    <span
      className="text value"
    >
      banana
    </span>
    <span
      className="faded text count"
    >
      155
    </span>
    <span
      className="faded text age"
    >
      100
    </span>
  </div>
  <div
    className="cardView rowView"
  >
    <span
      className="faded text number"
    >
      3
    </span>
    <span
      className="text value"
    >
      pear
    </span>
    <span
      className="faded text count"
    >
      17
    </span>
    <span
      className="faded text age"
    >
      5
    </span>
  </div>
</div>
`;

exports[`Renders a single RowView 1`] = `
<div
  className="cardView rowView"
>
  <span
    className="faded text number"
  >
    0
  </span>
  <span
    className="text value"
  >
    apple
  </span>
  <span
    className="faded text count"
  >
    7
  </span>
  <span
    className="faded text age"
  >
    4
  </span>
</div>
`;

exports[`Renders error DataTable 1`] = `
<div
  className="dataTable"
>
  <div
    className="headerView"
  >
    <span
      className="text number"
    >
      #
    </span>
    <span
      className="text value"
    >
      Value
    </span>
    <span
      className="text count"
    >
      Count
    </span>
    <span
      className="text age"
    >
      Age
    </span>
  </div>
  <div
    className="cardView errorView"
  >
    <span
      className="text"
    >
      Something went wrong.
    </span>
  </div>
  <div
    className="cardView reloadView"
    onClick={[Function]}
  >
    <span
      className="text"
    >
      Reload 
      <b>
        
      </b>
    </span>
  </div>
</div>
`;

exports[`Renders normal ReloadView. 1`] = `
<div
  className="cardView reloadView"
>
  <span
    className="text"
  >
    Reload 
    <b>
      
    </b>
  </span>
</div>
`;

exports[`Renders stale ReloadView. 1`] = `
<div
  className="cardView reloadView"
>
  <span
    className="text"
  >
    Reload 
    <b>
       - May be outdated.
    </b>
  </span>
</div>
`;
